src/metrics.rs

```diff
@@ -0,0 +1,87 @@
+use std::time::Instant;
+
+use axum::{http::Request, middleware::Next, response::IntoResponse};
+use lazy_static::lazy_static;
+use prometheus::{self, Encoder, HistogramOpts, HistogramVec, IntCounterVec, Opts};
+
+lazy_static! {
+    // Simple request counter
+    pub static ref INCOMING_REQUESTS: IntCounterVec = IntCounterVec::new(
+        Opts::new("incoming_requests", "The number of HTTP requests received"),
+        &["http_method", "path"]
+    ).expect("Prometheus metric initialization failed");
+    // Request counter by status code
+    pub static ref RESPONSE_CODE_COLLECTOR: IntCounterVec = IntCounterVec::new(
+        Opts::new("outgoing_response", "The number of responses sent"),
+        &["status_code", "http_method", "path"]
+    ).expect("Prometheus metric initialization failed");
+    // Response histogram by response time
+    pub static ref RESPONSE_TIME_COLLECTOR: HistogramVec = HistogramVec::new(
+        HistogramOpts{
+            common_opts: Opts::new("response_time", "The time taken to respond to each request"),
+            buckets: prometheus::DEFAULT_BUCKETS.to_vec(), // Change buckets here if desired
+        },
+        &["status_code", "http_method", "path"],
+    ).expect("Prometheus metric initialization failed");
+}
+
+/// Registers various prometheus metrics with the global registry
+pub fn register_metrics() {
+    let registry = prometheus::default_registry();
+    registry
+        .register(Box::new(INCOMING_REQUESTS.clone()))
+        .expect("registering prometheus metrics during initialization failed");
+    registry
+        .register(Box::new(RESPONSE_CODE_COLLECTOR.clone()))
+        .expect("registering prometheus metrics during initialization failed");
+    registry
+        .register(Box::new(RESPONSE_TIME_COLLECTOR.clone()))
+        .expect("registering prometheus metrics during initialization failed");
+}
+
+/// Returns currently gathered prometheus metrics
+pub async fn metrics_handler() -> String {
+    let encoder = prometheus::TextEncoder::new();
+    let mut buffer = Vec::new();
+
+    encoder
+        .encode(&prometheus::gather(), &mut buffer)
+        .expect("could not encode gathered metrics into temporary buffer");
+
+    let output =
+        String::from_utf8(buffer.clone()).expect("could not convert metrics buffer into string");
+    buffer.clear();
+
+    output
```

Comment:
I don't think we need the clone here. This is untested but compiles...

```suggestion
        String::from_utf8(buffer.clone()).expect("could not convert metrics buffer into string")
```

Reply:
Tested, seems to work.

Reply:
You mean the `clear` here right? Or am I just mis-reading the git diff...?

Reply:
I meant the clone, but forgot to drop it...

If you remove the clone you also need to remove the (unnecessary) clear as well to keep the borrow checker happy.

Reply:
Ah okay, that makes sense - good spot!